### PR TITLE
Don't overload the `bool()` intrinsic function

### LINKED ISSUE #410: Don't overload the `bool` intrinsic function
When I run this:
```python
def f():
  i: i32
  i = 3
  b: bool
  b = bool(i)
  print(b)


f()
```
```console
% lpython --show-cpp c.py
#include <iostream>
#include <string>
#include <vector>
#include <cassert>
#include <cmath>
#include <complex>
#include <Kokkos_Core.hpp>
#include <lfortran_intrinsics.h>

template <typename T>
Kokkos::View<T*> from_std_vector(const std::vector<T> &v)
{
    Kokkos::View<T*> r("r", v.size());
    for (size_t i=0; i < v.size(); i++) {
        r(i) = v[i];
    }
    return r;
}

void _lpython_main_program()
{
    f();
}

void f()
{
    bool b;
    int i;
    i = 3;
    b = __lpython_overloaded_0__bool(i);
    std::cout << b << std::endl;
}

bool __lpython_overloaded_0__bool(int x)
{
    bool _lpython_return_variable;
    _lpython_return_variable = x != 0;
    return _lpython_return_variable;
}

namespace {

void main2() {
        _lpython_main_program();
}

}
int main(int argc, char* argv[])
{
    Kokkos::initialize(argc, argv);
    main2();
    Kokkos::finalize();
    return 0;
}

```
I think in this case, `f()` function could be as simple as:
```c++
void f()
{
    bool b;
    int i;
    i = 3;
    b = (bool)i;
    std::cout << b << std::endl;
}
```
That can happen if we don't overload the `bool` intrinsic function, and use `IntegerToLogical` casting in AST->ASR.
So, in other words, all we have to do is implement `handle_intrinsic_bool`, similar to

https://github.com/lcompilers/lpython/blob/2739a8596a4bad5528a75f4aa290af4e7b15443d/src/lpython/semantics/python_ast_to_asr.cpp#L2572

Comment out the current implementation of `bool` in `lpython_builtin.py` and `python_comptime_eval.h`,

And modify the C++ backend (`Cast` visitor):
```diff
diff --git a/src/libasr/codegen/asr_to_cpp.cpp b/src/libasr/codegen/asr_to_cpp.cpp
index 9dd4732cd..3af93c340 100644
--- a/src/libasr/codegen/asr_to_cpp.cpp
+++ b/src/libasr/codegen/asr_to_cpp.cpp
@@ -657,6 +657,10 @@ Kokkos::View<T*> from_std_vector(const std::vector<T> &v)
                 src = "(int)(" + src + ")";
                 break;
             }
+            case (ASR::cast_kindType::IntegerToLogical) : {
+                src = "(bool)(" + src + ")";
+                break;
+            }
             default : throw CodeGenError("Cast kind " + std::to_string(x.m_kind) + " not implemented");
         }
         last_expr_precedence = 2
```

### PR DESCRIPTION
**Pull request by namannimmo10**
Fixes #410

### CODE REVIEW

**Comment by certik** [comment]
Something like that.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -3452,6 +3452,72 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+                al, loc, t, ASR::cast_kindType::IntegerToLogical, to_type, value));
+
+        } else if (ASRUtils::is_character(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                char *c = ASR::down_cast<ASR::StringConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_s;
+                value = ASR::down_cast<ASR::expr_t>(make_LogicalConstant_t(al,
+                                loc, std::string(c) != "", to_type));
+            }
+            // TODO: make bool() work for non-constant strings
```
Comment (namannimmo10) [inline src/lpython/semantics/python_ast_to_asr.cpp:3500]: How do you wanna do this? Call a runtime function or have a node for this, too?

Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp:3500]: Can this be done with the existing `Cast` ASR node? Such as casting Integer to Bool.

Comment (czgdp1807) [reply src/lpython/semantics/python_ast_to_asr.cpp:3500]: I would suggest using `Cast` node. What do you all think?

Comment (namannimmo10) [reply src/lpython/semantics/python_ast_to_asr.cpp:3500]: How about we add `CharacterToLogical` and `ComplexToLogical` under `cast_kind`? 

> Can this be done with the existing `Cast` ASR node?

I don't see any existing node that does the job here.

Comment (czgdp1807) [reply src/lpython/semantics/python_ast_to_asr.cpp:3500]: Well here’s what I think.
1. First check whether the call is being made to `bool`. In an abstract sense, I mean verify that the semantics of the code is to call `bool` over a single input.
2. Now, see if the input is of intrinsic type (i.e., integer, real/float, complex). If so then things are pretty much clear, source type is the type of input argument and destination type is logical. If the input type is say a class then I think Python provides some magic method (I think it is `__bool__`) for conversion. We will have to call `__bool__`.
3. On the basis of step 2 either create a `Cast` node (if the type is node) or a `FunctionCall` (if the type is class).

I think I have covered all the cases for `bool(x)` expression. It is fairly easy to implement the above steps in ASR (provided I am not missing out on anything).

Comment (czgdp1807) [reply src/lpython/semantics/python_ast_to_asr.cpp:3500]: > How about we add CharacterToLogical and ComplexToLogical under cast_kind?

I think yes we should add and use them in step 3 of my comment above.

Comment (czgdp1807) [reply src/lpython/semantics/python_ast_to_asr.cpp:3500]: For non-constant strings in the LLVM backend we might not need a new C runtime function. We can call already existing function which computes the length of the string. The integer received in return will easily determine the result. If 0 then false and true in all other cases.

**Comment by namannimmo10** [comment]
@czgdp1807, @certik -- this is ready.

**Review by certik — APPROVED** [review]
I think that this is fine.

**Pull request merged**